File: src/Repository.ts
import { SqlClient } from "@effect/sql"
import * as Effect from "effect/Effect"
import * as Schema from "effect/Schema"

import { Artifact, type Slug } from "./Domain.js"

export interface NewArtifact {
  readonly id: string
  readonly slug: string
  readonly title: string
  readonly description: string | null
  readonly sourceType: string
  readonly sourceFilename: string
  readonly sourcePath: string
  readonly sha256: string
  readonly sizeBytes: number
  readonly project: string | null
  readonly repoFullName: string | null
  readonly branch: string | null
  readonly commitSha: string | null
  readonly dirty: boolean
  readonly agent: string | null
  readonly generator: string | null
  readonly state: "active"
  readonly createdAt: string
  readonly updatedAt: string
}

interface ArtifactRow {
  readonly id: string
  readonly slug: string
  readonly title: string
  readonly description: string | null
  readonly source_type: string
  readonly source_filename: string
  readonly source_path: string
  readonly sha256: string
  readonly size_bytes: number
  readonly project: string | null
  readonly repo_full_name: string | null
  readonly branch: string | null
  readonly commit_sha: string | null
  readonly dirty: number
  readonly agent: string | null
  readonly generator: string | null
  readonly state: string
  readonly created_at: string
  readonly updated_at: string
}

const rowToArtifact = (row: ArtifactRow) =>
  Schema.decodeUnknownSync(Artifact)({
    id: row.id,
    slug: row.slug,
    title: row.title,
    description: row.description,
    sourceType: row.source_type,
    sourceFilename: row.source_filename,
    sourcePath: row.source_path,
    sha256: row.sha256,
    sizeBytes: row.size_bytes,
    project: row.project,
    repoFullName: row.repo_full_name,
    branch: row.branch,
    commitSha: row.commit_sha,
    dirty: row.dirty === 1,
    agent: row.agent,
    generator: row.generator,
    state: row.state,
    createdAt: row.created_at,
    updatedAt: row.updated_at
  })

export const initializeDatabase = Effect.gen(function*() {
  const sql = yield* SqlClient.SqlClient
  yield* sql.unsafe(`
    create table if not exists artifacts (
      id text primary key,
      slug text not null unique,
      title text not null,
      description text,
      source_type text not null,
      source_filename text not null,
      source_path text not null,
      sha256 text not null,
      size_bytes integer not null,
      project text,
      repo_full_name text,
      branch text,
      commit_sha text,
      dirty integer not null default 0,
      agent text,
      generator text,
      state text not null default 'active',
      created_at text not null,
      updated_at text not null
    )
  `)
  yield* sql.unsafe("create index if not exists artifacts_created_at_idx on artifacts(created_at desc)")
  yield* sql.unsafe("create index if not exists artifacts_state_idx on artifacts(state)")
})

export const insertArtifact = (artifact: NewArtifact) =>
  Effect.gen(function*() {
    const sql = yield* SqlClient.SqlClient
    yield* sql`
      insert into artifacts ${
      sql.insert({
        id: artifact.id,
        slug: artifact.slug,
        title: artifact.title,
        description: artifact.description,
        source_type: artifact.sourceType,
        source_filename: artifact.sourceFilename,
        source_path: artifact.sourcePath,
        sha256: artifact.sha256,
        size_bytes: artifact.sizeBytes,
        project: artifact.project,
        repo_full_name: artifact.repoFullName,
        branch: artifact.branch,
        commit_sha: artifact.commitSha,
        dirty: artifact.dirty ? 1 : 0,
        agent: artifact.agent,
        generator: artifact.generator,
        state: artifact.state,
        created_at: artifact.createdAt,
        updated_at: artifact.updatedAt
      })
    }
    `
  })

export const findArtifactBySlug = (slug: Slug) =>
  Effect.gen(function*() {
    const sql = yield* SqlClient.SqlClient
    const rows = yield* sql<ArtifactRow>`select * from artifacts where slug = ${slug} limit 1`
    return rows[0] === undefined ? null : rowToArtifact(rows[0])
  })

export const slugExists = (slug: Slug) =>
  Effect.gen(function*() {
    const sql = yield* SqlClient.SqlClient
    const rows = yield* sql<{ readonly count: number }>`select count(*) as count from artifacts where slug = ${slug}`
    return (rows[0]?.count ?? 0) > 0
  })

export const listRecentArtifacts = (limit: number) =>
  Effect.gen(function*() {
    const sql = yield* SqlClient.SqlClient
    const rows = yield* sql<ArtifactRow>`
      select * from artifacts
      where state = 'active'
      order by created_at desc
      limit ${limit}
    `
    return rows.map(rowToArtifact)
  })

File: internal/modules/artifacts/repository.go
package artifacts

import (
	"context"
	"errors"
	"fmt"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgconn"
	"github.com/jackc/pgx/v5/pgtype"
	"github.com/mholtzscher/agent-artifacts/internal/platform/db/sqlc"
)

type Repository interface {
	Create(context.Context, Artifact) error
	FindBySlug(context.Context, string) (Artifact, error)
	ListRecent(context.Context, int32) ([]Artifact, error)
}

type SQLRepository struct{ queries *sqlc.Queries }

func NewSQLRepository(queries *sqlc.Queries) *SQLRepository { return &SQLRepository{queries: queries} }
func (r *SQLRepository) Create(ctx context.Context, value Artifact) error {
	err := r.queries.CreateArtifact(ctx, createArtifactParams(value))
	if isSlugCollision(err) {
		return ErrSlugTaken
	}
	return err
}
func (r *SQLRepository) FindBySlug(ctx context.Context, slug string) (Artifact, error) {
	row, err := r.queries.GetArtifactBySlug(ctx, sqlc.GetArtifactBySlugParams{Slug: slug})
	if errors.Is(err, pgx.ErrNoRows) {
		return Artifact{}, ErrNotFound
	}
	if err != nil {
		return Artifact{}, fmt.Errorf("find artifact: %w", err)
	}
	return artifactFromRow(row)
}
func (r *SQLRepository) ListRecent(ctx context.Context, limit int32) ([]Artifact, error) {
	rows, err := r.queries.ListRecentArtifacts(ctx, sqlc.ListRecentArtifactsParams{Limit: limit})
	if err != nil {
		return nil, fmt.Errorf("list recent artifacts: %w", err)
	}
	values := make([]Artifact, 0, len(rows))
	for _, row := range rows {
		value, err := artifactFromRow(row)
		if err != nil {
			return nil, fmt.Errorf("decode recent artifact: %w", err)
		}
		values = append(values, value)
	}
	return values, nil
}
func createArtifactParams(value Artifact) sqlc.CreateArtifactParams {
	return sqlc.CreateArtifactParams{ID: pgtype.UUID{Bytes: [16]byte(value.ID), Valid: true}, Slug: value.Slug, Title: value.Title, Description: nullableText(value.Description), SourceType: sqlc.SourceType(value.SourceType), SourceFilename: value.SourceFilename, Sha256: value.SHA256, SizeBytes: value.SizeBytes, Project: nullableText(value.Project), RepoFullName: nullableText(value.RepoFullName), Branch: nullableText(value.Branch), CommitSha: nullableText(value.CommitSHA), Dirty: value.Dirty, Agent: nullableText(value.Agent), Generator: nullableText(value.Generator), CreatedAt: pgtype.Timestamptz{Time: value.CreatedAt, Valid: true}, UpdatedAt: pgtype.Timestamptz{Time: value.UpdatedAt, Valid: true}}
}
func artifactFromRow(row sqlc.Artifact) (Artifact, error) {
	if !row.ID.Valid || !row.CreatedAt.Valid || !row.UpdatedAt.Valid {
		return Artifact{}, fmt.Errorf("artifact row contains null required values")
	}
	return Artifact{ID: uuid.UUID(row.ID.Bytes), Slug: row.Slug, Title: row.Title, Description: textPointer(row.Description), SourceType: SourceType(row.SourceType), SourceFilename: row.SourceFilename, SHA256: row.Sha256, SizeBytes: row.SizeBytes, Project: textPointer(row.Project), RepoFullName: textPointer(row.RepoFullName), Branch: textPointer(row.Branch), CommitSHA: textPointer(row.CommitSha), Dirty: row.Dirty, Agent: textPointer(row.Agent), Generator: textPointer(row.Generator), CreatedAt: row.CreatedAt.Time, UpdatedAt: row.UpdatedAt.Time}, nil
}
func nullableText(value *string) pgtype.Text {
	if value == nil {
		return pgtype.Text{}
	}
	return pgtype.Text{String: *value, Valid: true}
}
func textPointer(value pgtype.Text) *string {
	if !value.Valid {
		return nil
	}
	copy := value.String
	return &copy
}
func isSlugCollision(err error) bool {
	var postgresError *pgconn.PgError
	return errors.As(err, &postgresError) && postgresError.Code == "23505" && postgresError.ConstraintName == "artifacts_slug_key"
}
